File: src/net/liplum/blocks/StatedWall.java
package net.liplum.blocks;

import arc.Core;
import arc.graphics.g2d.Draw;
import arc.graphics.g2d.TextureRegion;
import mindustry.graphics.Drawf;
import mindustry.world.blocks.defense.Wall;
import net.liplum.components.ComponentBase;

import java.util.ArrayList;

public class StatedWall extends Wall {
    public TextureRegion[] states;
    public float shadowRadius = 20f;
    public int stateNumber;
    public ArrayList<ComponentBase<StatedWallBuild>> components =
            new ArrayList<>();

    public StatedWall(String name) {
        super(name);
    }

    @Override
    public void load() {
        super.load();
        states = new TextureRegion[stateNumber];
        for (int i = 0; i < stateNumber; i++) {
            states[i] = Core.atlas.find(name + "-" + i);
        }
    }

    public class StatedWallBuild extends WallBuild {
        @Override
        public void updateTile() {
            for (ComponentBase<StatedWallBuild> component : components) {
                component.onUpdate(this);
            }
        }

        @Override
        public void drawCracks() {
        }

        @Override
        public void draw() {
            if (shadowRadius > 0f) {
                Drawf.shadow(x, y, shadowRadius);
            }
            int curIndex = (int) (lostHealthPct() * stateNumber);
            curIndex = Math.min(curIndex, stateNumber - 1);
            Draw.rect(states[curIndex], x, y);
            this.drawTeamTop();
        }

        public float lostHealthPct() {
            return 1f - health / maxHealth;
        }
    }
}

File: src/net/liplum/blocks/StatedWall.kt
package net.liplum.blocks

import arc.Core
import arc.graphics.g2d.Draw.*
import arc.graphics.g2d.TextureRegion
import mindustry.graphics.Drawf
import mindustry.world.blocks.defense.Wall
import net.liplum.components.ComponentBase

open class StatedWall(name: String) : Wall(name) {
    @JvmField var states: Array<TextureRegion> = arrayOf()
    @JvmField var shadowRadius = 20f
    @JvmField var stateNumber = 0
    @JvmField var components = ArrayList<ComponentBase<StatedWallBuild>>()
    @JvmField var drawSize = 1f
    override fun load() {
        super.load()
        states = Array(stateNumber) {
            Core.atlas.find("$name-$it")
        }
    }

    inner class StatedWallBuild : WallBuild() {
        override fun updateTile() {
            for (component in components) {
                component.onUpdate(this)
            }
        }

        override fun drawCracks() {}
        override fun draw() {
            if (shadowRadius > 0f) {
                Drawf.shadow(x, y, shadowRadius)
            }
            var curIndex = (lostHealthPct * stateNumber).toInt()
            curIndex = curIndex.coerceAtMost(stateNumber - 1)
            rect(
                states[curIndex], x, y,
                region.width * scl * xscl * drawSize,
                region.height * scl * yscl * drawSize
            )
            drawTeamTop()
        }

        val lostHealthPct: Float
            get() {
                return 1f - health / maxHealth
            }
    }
}